### PR TITLE
Add docs to a pytorch.github.io/doc/tag directory when repo is tagged

### PR DESCRIPTION
**Pull request by mattip**
In coordination with @jlin27.

This PR is meant to build documentation when the repo is tagged. For instance, tagging the repo with 1.7.0rc1 will push that commit's documentation to pytorch/pytorch.github.io/docs/1.7. 

Subsequently tagging 1.7.0rc2 will override the 1.7 docs, as will 1.7.0, and 1.7.1. I think this is as it should be: there should be one, latest, version for the 1.7 docs. This can be tweaked differently if desired.

There is probably work that needs to be done to adjust the [versions.html](https://pytorch.org/docs/versions.html) to add the new tag?

Is there a way to test the tagging side of this without breaking the production documentation?

As an aside, the documentation is being built via the `pytorch_linux_xenial_py3_6_gcc5_4_build` image. Some projects are starting to move on from python3.6 since [it is in security-only support mode](https://devguide.python.org/#status-of-python-branches), no new binaries are being released.

### CODE REVIEW

**Comment by mattip** [comment]
@seemethere the build is successful when triggered on master. Is there a way to do any kind of test for a new tag?

File: .circleci/cimodel/data/pytorch_build_definitions.py [hunk 1]
```diff
@@ -348,6 +349,8 @@ def instantiate_configs():
 
         # run docs builds on "pytorch-linux-xenial-py3.6-gcc5.4". Docs builds
         # should run on a CPU-only build that runs on all PRs.
+        # XXX should this be updated to a more modern build? Projects are
+        #     beginning to drop python3.6
```
Comment (zou3519) [inline .circleci/cimodel/data/pytorch_build_definitions.py:353]: I think it's reasonable to move this to a more modern build. We could do that in a follow-up PR

**Comment by zou3519** [comment]
> Subsequently tagging 1.7.0rc2 will override the 1.7 docs, as will 1.7.0, and 1.7.1. I think this is as it should be: there should be one, latest, version for the 1.7 docs. This can be tweaked differently if desired.

In the past, we've kept around the documentation for dot releases separately. 

image

> There is probably work that needs to be done to adjust the versions.html to add the new tag?

Probably. Some work also needs to be done in moving the stable/ symlink to point to the right thing when we do releases. I think it's ok if we don't automate those two things for now (it's very easy to just toss up a PR to pytorch.github.io to adjust versions.html and the stable symlink)

**Comment by mattip** [comment]
> In the past, we've kept around the documentation for dot releases separately.

Changed in 9ed16dc  to keep this behaviour. It would be nice to merge this before the upcoming release cycle.

**Comment by zou3519** [comment]
This looks fine to me but I don't know much about the circleci parts. @seemethere could you take a look please?

File: .circleci/cimodel/data/pytorch_build_definitions.py [hunk 2]
```diff
@@ -166,7 +166,8 @@ def gen_workflow_job(self, phase):
                 "branch": self.branch,
                 "requires": [self.parent_build],
                 "context": "org-member",
-                "filters": gen_filter_dict(branches_list=["nightly"])
+                "filters": gen_filter_dict(branches_list=["nightly"],
+                                           tags_list=RC_PATTERN)
```
Comment (seemethere) [inline .circleci/cimodel/data/pytorch_build_definitions.py:170]: On gotcha that I can foresee potentially happening is that CircleCI has a weird requirement that not all of the jobs can run on a tag, so we'd not only have to add this `gen_filter_dict` call here but we'd need to add it for all of the parent jobs as well. (i.e. all of the jobs that the docs builds `requires`)

**Review by seemethere — COMMENTED** [review]
overall LGTM, there is one gotcha though that might limit the effectiveness on release day.

File: .circleci/cimodel/data/pytorch_build_definitions.py (hunk 2 above)
Comment (mattip) [reply .circleci/cimodel/data/pytorch_build_definitions.py:170]: Ahh, I thought that would resolve automatically. The parent jobs are `pytorch_cpp_doc_build`, `pytorch_python_doc_build`, and the `pytorch-linux-xenial-py3.6-gcc5.4` composite job.

**Comment by mattip** [comment]
I added tag filters to all the parents of the doc push jobs. The one in docker_definitions.py is particularly ugly, but I think it works.

**Review by seemethere — APPROVED** [review]

**Comment by seemethere** [comment]
Thanks for the contribution @mattip! 😄

**Comment by mattip** [comment]
@seemethere it seems something is not quite right with this pr. 1.7.0rc1 was tagged, and the [build triggered](https://app.circleci.com/pipelines/github/pytorch/pytorch/221556/workflows/410e84cd-d0e5-45e4-b6a6-0f8e775ce301), but bothe the [pytorch_cpp_doc_push](https://app.circleci.com/pipelines/github/pytorch/pytorch/221556/workflows/410e84cd-d0e5-45e4-b6a6-0f8e775ce301/jobs/7903100) and the [pytorch_python_doc_push](https://app.circleci.com/pipelines/github/pytorch/pytorch/221556/workflows/410e84cd-d0e5-45e4-b6a6-0f8e775ce301/jobs/7902524) failed. They are missing the `$GITHUB_PYTORCHBOT_TOKEN` ?

**Comment by mattip** [comment]
Now the doc push is running, but [it fails](https://app.circleci.com/pipelines/github/pytorch/pytorch/221773/workflows/e832ecfd-09c7-46b9-a4e7-e17fb8652aa1/jobs/7916599).

<details>

```
#!/bin/bash -eo pipefail
pushd /tmp/workspace
git push -u origin "site"

/tmp/workspace ~/project
To https://github.com/pytorch/pytorch.github.io
 ! [rejected]                site -> site (fetch first)
error: failed to push some refs to 'https://github.com/pytorch/pytorch.github.io'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.

Exited with code exit status 1

CircleCI received exit code 1
```

**Comment by mattip** [comment]
Not sure how, but something created the docs https://pytorch.org/docs/1.7.0/. Was this PR not needed, or did it succeed in spite of the run I pointed to failing?

**Comment by seemethere** [comment]
Oh the first one was an abort, we had to re-run due to an issue with CircleCI

**Pull request closed without merging**